### ISSUE TITLE
Introduce Selection class.

### ISSUE THREAD

**Issue by scofalik** [ckeditor/ckeditor5-core#113]
Introduce `Selection` class. Some requirements:
* should support multiple ranges
* should support `anchor` and `focus` positions
* should updates itself automatically when operation is applied (this will be done thanks to `LiveRange`)
* should ~~be~~ *somehow* ~~connected with `Batch` to~~ support undo *(will figure it out later)*
* should be a `treeModel.Document` property
* should emit events but only when changed by one of it's methods (no events emitting when it gets changed by operations)

**Comment by scofalik** [comment]
* should convert `Range` to `LiveRange`

Ideas:
* should ranges kept in `Selection` instance be sorted?
* when range is added and it intersects with other range that is already added, should those ranges get merged?

**Comment by oleq** [comment]
> should support multiple ranges
AFAIR it's only supported in Firefox and we actually neglected it in many places in the course of v4 development. So, why bother?

**Comment by scofalik** [comment]
That was F2F decision we made. Also, how we will handle table selections?

**Comment by Reinmar** [comment]
There should be support for multi-range selections because we need to be able to at least "fake" columns selections. It does not have to handled natively for us to handle it in the API.

**Comment by scofalik** [comment]
Explanation: `Selection` probably should not be connected with `Batch` itself but with undo system. This may get figured out when we will implement undo.

**Comment by Reinmar** [comment]
> should ranges kept in Selection instance be sorted?

I don't think so.

> when range is added and it intersects with other range that is already added, should those ranges get merged?

Throw! (I'm not joking :P). It's a similar case as with inverted ranges – to avoid more serious issues it's better to reject such situation.

**Comment by scofalik** [comment]
> Throw! (I'm not joking :P). It's a similar case as with inverted ranges – to avoid more serious issues it's better to reject such situation.

Another solution is to add only a difference of added range and stored ranges to the set of already added ranges.

**Comment by pjasiun** [comment]
> we should not expect user to detach all LiveRanges before using .removeAllRanges() or other API methods that removes ranges (if they will be implemented).

I think user should not need to care about detaching, or know how live ranges work, at all. Like in  the DOM: you set the range and it is in the (more-less) good place, even if you change a DOM structure a little. You do not need to know how it works.

> > should ranges kept in Selection instance be sorted?
>
> I don't think so.

Nighter do I.

> > when range is added and it intersects with other range that is already added, should those ranges get merged?
>
> Throw! (I'm not joking :P).

We can not think about ranges as we were thinking about them in CKEditor 4. Because of OT strange things can happen. In this case editor should not throw errors when one user have a multi-selection and another move part of the DOM. Merging them sounds reasonable.

Anyway I believe it is an edge case. Whole multi-selection is an edge case at this stage, but we need it, at least for tables.

> should emit events when changed by user in editable?

The selection in the editable is a different selection then this one, in the model. `EditableView` will fire an event then the selection in the editable changes. This event should be handled by a controller and causes selection change on the model. This change will be done the same way as features change the selection, using the selection API. Then such change should fire the event so the EditableView will update selection property and set the selection in the proper position: keep it in the position it was, or move to the better place.

**Comment by scofalik** [comment]
> I think user should not need to care about detaching, or know how live ranges work, at all. Like in the DOM: you set the range and it is in the (more-less) good place, even if you change a DOM structure a little. You do not need to know how it works.

It's nice in browser when you can use a language that supports destructors or other way to handle some things. But unfortunately right now we will need to have a `.detach()` party in the code.

**Comment by scofalik** [comment]
I've pushed commit with tests for `Selection` class so you may look at it's API and behavior and comment. I'd really like to hear what you think before I start coding those methods.

Few things that came to my mind:
- I don't like that `LiveRange` is stored as is and `Range` is going to be converted into `LiveRange` (so a new object is created, that lives only inside `Selection` instance - contrary to `LiveRange` that lives outside `Selection` instance).
- This is problematic for `Selection.detach()`. It should unbind all events but when `LiveRange` is passed, `Selection` is not making a copy of it - but still detaches it. So you might detach a `LiveRange` that user may expect to still have events bound.
- This leads me to an idea that `Selection` should always make a copy of added range inside it.
- Or `Selection.addRange` could take `Position`s as parameters and make `LiveRange` inside `Selection`. This would also solve "backward/forward" selection problem - right now it is solved by passing an additional flag with a range when calling `Selection.addRange`. Unfortunately this would make some parts of code longer (`sel.addRange( myRange )` vs `sel.addRange( myRange.start, myRange.end )`).
- `Selection.getRanges` probably should return copies of ranges stored in selection, so user won't be able to operate on them outside of `Selection` instance.
- I was thinking about introducing `Selection.getCombinedRange` (or simply `.getRange`). It would work like `sel.getRanges()[ 0 ]` if selection has only one range but would be useful in other cases: it would return `null` if selection has no ranges and would return one big range if selection has more ranges (the range would be a minimal range that contains all ranges added to selection). Question is whether we need that in API.

**Comment by pjasiun** [comment]
> I don't like that LiveRange is stored as is and Range is going to be converted into LiveRange (so a new object is created, that lives only inside Selection instance - contrary to LiveRange that lives outside Selection instance).
> This is problematic for Selection.detach(). It should unbind all events but when LiveRange is passed, Selection is not making a copy of it - but still detaches it. So you might detach a LiveRange that user may expect to still have events bound.
> This leads me to an idea that Selection should always make a copy of added range inside it.

I agree that selection should always create a it's own copy of the `LiveRange`. So if user create a selection based on the live range and he detach this range later he will not break selection. And if the selection will detach range it will not break that range. Otherwise no one will know if it can detach range.

**Comment by scofalik** [comment]
Pushed https://github.com/ckeditor/ckeditor5-core/tree/t/113-b

I updated tests a bit. `Selection` now always creates it's own `LiveRange`s and in `.detach()` it detaches events bound by those `LiveRange`s.

The branch is now based on t/99-e ( see for more details #64 ).

To do:
* Inserting intersecting ranges.
* Decide whether `.getRanges()` should clone ranges that are added to selection.

**Comment by scofalik** [comment]
For now I decided that intersecting range will throw an error when added to the selection. If this will become a problem, it will be changed. This solution is easier to implement and, in fact, may help us find some bugs.

I also introduced `Range.isIntersecting`. I know this method already existed and got removed and we agreed, that one can use `if ( a.getIntersection( b ) !== null )` to check if ranges intersect but:
1) Right now this method will be used in three places and I expect that it will be heavily used by API users.
2) It is very simple and keeping it in API won't cost us anything. In worst case scenario we will change it to `return this.getIntersection( otherRange ) !== null` as described above.
3) If ranges intersect, `Range.getIntersection()` calls (in total) four methods that compare two arrays and then creates new object, which won't be used anywhere.

I think that this soultion is cleaner and has better performance and does not cost us anything.

For now, if you don't have any other ideas or complaints, I think this is PR-ready. Unfortunately it should wait for #64 .
